doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+
+/* ... */
+
+/* Dummy implementation of `get` handler.
+   For both configuration parameters, it copies the value to a `val` variable.
+*/
+int my_get_handler(int argc, char **argv, char *val, int val_len_max)
+{
+    if (argc) {
+        if (!strcmp("is_enabled", argv[0])) {
+            /* Copy the value of `is_enabled` to `val` so the user can read it */
+            memcpy(val, &is_enabled, sizeof(is_enabled));
+        }
+        else if (!strcmp("threshold", argv[0])) {
+            /* Copy the value of `is_enabled` to `val` so the user can read it */
+            memcpy(val, &threshold, sizeof(threshold));
+        }
+    }
+    /* ... */
+    return NULL;
+}
+/* Dummy implementation of `set` handler.
+   For both configuration parameters, it sets the value from `val`.
+*/
+int my_set_handler(int argc, char **argv, char *val)
+{
+    if (argc) {
+        if (!strcmp("is_enabled", argv[0])) {
+            /* Set the value of `is_enabled` from `val` */
+            memcpy(&is_enabled, val, sizeof(is_enabled));
+        }
+        else if (!strcmp("threshold", argv[0])) {
+            /* Validate threshold */
+            if(atoi(val) > MAX_THRESHOLD)
+                return -EINVAL;
+
+            /* Set the value of `theshold` from `val` */
+            memcpy(&threshold, val, sizeof(threshold));
+        }
+    }
+    /* ... */
+    return 0;
+}
+
+/* Dummy implementation of `commit` handler.
+   This is intended to be called by the Registry when all configurations have
+   to be applied. Because of this, it's possible to implement transactions or
+   protect against faulty combinations of configs, race conditions, etc.
+*/
+int my_commit_handler(int argc, char **argv, char *val)
+{
+    /* Do something if the module is enable */
+    if(is_enabled) {
+        trigger_something();
+    }
+    
+    /* As stated before, the application crashes if `set_threashold` is called when is_enabled is false.
+    We protect it here */
+    if(is_enabled) {
+        /* We can safely set the theshold without crashing the app */
+        set_threshold(threshold);
+    }
+}
+
+/* Dummy implementation of `export` handler.
+   This handler has to call `export_func` for one or all parameters.
+   There can be different behaviors depending on the export function (e.g printing all configs
+   to STDOUT, save them in a non-volatile storage device, etc)
+*/
+void my_export_handler(int (*export_func)(const char *name, char *val), int argc,
+                   char **argv)
+{
+    /* argc  and argv can be used to export only one parameter */
+    (void)argv;
+    (void)argc;
+    char buf[INT_STRING_SIZE];
+
+    /* We export every parameter with the export function */
+    
+    /* Prepare `buf` to contain is_enabled in a string representation */
+    /* ... */
+    export_func("my_handler/is_enabled", buf);
+    /* Prepare `buf` to contain threshold in a string representation */
+    /* ... */
+    export_func("my_handler/threshold", buf);
```

Comment:
this is stated in the comments slightly above the export_func call

Reply:
Sorry don't understand - export_func loads `threshold` or `is_enabled` into `buf`, right? So the second `export_func` call would overwrite the contents of `buf`, and the first call wouldn't be useful?

Reply:
From the RDM:
> export: Calls an export function for each configuration parameter, with its name and value. Depending on the behavior of export function, this can be used for printing out all configurations, saving them in a persistent storage, etc.

`export_func` receives (key, value) and does whatever it has to do with those. In practice, it's used for printing all variables in STDOUT or storing them

Reply:
Ah! the comments refer to stuff that happens in between the calls, not to the functions themselves.
Maybe a space between the comments and the function calls?

Reply:
ok, I will rephrase that snippet